File: articles/linear-algebra-for-systems-modeling/articles/rank-nullity-and-structural-dependence/julia/rank_nullity_audit.jl
A = [
    1.0 1.0 0.0;
    0.0 1.0 1.0;
    1.0 0.0 1.0
]

row_count = size(A, 1)
column_count = size(A, 2)
rank_value = 3
nullity_value = column_count - rank_value
rank_deficient = rank_value < min(row_count, column_count)

println("system_name,row_count,column_count,rank,nullity,rank_deficient,pivot_columns,free_columns,warning")
println(join((
    "three_constraint_resource_balance_matrix",
    row_count,
    column_count,
    rank_value,
    nullity_value,
    rank_deficient,
    "0;1;2",
    "none",
    "Rank and nullity reveal structure but interpretation depends on model meaning."
), ","))

File: articles/linear-algebra-for-systems-modeling/articles/rank-nullity-and-structural-dependence/c/rank_nullity_audit.c
#include <stdio.h>

int main(void) {
    int row_count = 3;
    int column_count = 3;
    int rank = 3;
    int nullity = column_count - rank;
    int rank_deficient = 0;
    double tolerance = 1e-10;

    printf("system_name,row_count,column_count,rank,nullity,rank_deficient,pivot_columns,free_columns,tolerance,warning\n");
    printf("three_constraint_resource_balance_matrix,%d,%d,%d,%d,%d,0;1;2,none,%.10f,Rank and nullity reveal structure but interpretation depends on model meaning.\n",
           row_count, column_count, rank, nullity, rank_deficient, tolerance);
    return 0;
}
